### ISSUE TITLE
Test performance degradation

### ISSUE THREAD

**Issue by rvernica** [Paradigm4/shim#111]
The last test suite in Shim is a multi-session load test. It starts 10 threads and maxes out the number of Shim sessions. Each session creates an array, runs an apply/cross_join query, and removes the array. With SciDB `19.11`, the test suite takes too long and Travis stops the show after 10 minutes. See https://www.travis-ci.org/github/Paradigm4/shim/builds/667959804#L2209 This test takes ~5 min in Travis with SciDB `19.3`

The workaround is to reduce number of threads to two (`-n 2`) for the test to finish in time before the Travis timeout.

It has been confirmed that on a more powerful servers than the one used in Travis, the test does complete successfully after more than 10 minutes.

**Comment by rvernica** [comment]
It seems that this issue is triggered by the `remove` operator. If the test suite is run [without](https://github.com/Paradigm4/shim/blob/549dc10f7f92eba356f82beaea675c7a224deec3/tests/pytest/test_sessions_multi.py#L66) the `remove` statements it works fine. 

The full test suite with 10 parallel threads completes successfully in Travis. Instead of timing-out after 10 minutes due to no output, the suite completes in approximately five minutes.

A similar behavior was observed in a different context in https://github.com/rvernica/docker-library/issues/12

**Comment by ksens** [comment]
@rvernica Super! Thanks for getting to the bottom of this. 

So what was the final thread setting that you used -- I am guessing you made a change to the scidb config file, right?

**Comment by dgosselin-p4** [comment]
@rvernica , @ksens 
I looked at a shim slowdown a couple of months ago.  I fixed a `remove()` performance problem in 19.11, but not in 19.3.  Interestingly, it appears here that 19.11 is slow during `remove()` while 19.3 is not.  I can investigate here locally.  Perhaps there's a problem with `remove()` in multiuser test cases?

**Comment by ksens** [comment]
Will let @rvernica respond. He has the full details.

**Comment by rvernica** [comment]
Yes, it appears to be a problem with `remove` in the multi-user test case while the system is loaded.

@ksens No changes were made to the SciDB config. The test is multi-threaded. On Travis, without `remove` it works with 10 threads. With `remove` it only works with 2 threads.

**Comment by ksens** [comment]
Just to confirm:

By threads, do you mean concurrent users / clients? Maybe in your multi-user test case, you spawn threads that send queries to shim?

**Comment by rvernica** [comment]
Correct. The test spawns 10 parallel threads. Each thread starts five Shim sessions. So, together they max out the number of sessions available in Shim (50). In each session, the test issues one `store`/`build` [query](https://github.com/Paradigm4/shim/blob/master/tests/pytest/test_sessions_multi.py#L14).

**Comment by dgosselin-p4** [comment]
Regardless of whether `remove()` is invoked, running `pytest -n10 test_sessions_multi.py` takes about 13 minutes to complete on my dev system.

**Comment by rvernica** [comment]
This is consistent with what you observed in the past. On my dev system, I can't go above `-n3`. I think it has to do with the overall system capacity. So, could you try `-n50` on your system? If this does not work, I can start a small instance in AWS and try to reproduce what I'm seeing.

**Comment by dgosselin-p4** [comment]
Regardless of system capacity, I would expect to see some change in the performance characteristics of the tests when the arrays were removed by the script versus when they're not removed.

**Comment by dgosselin-p4** [comment]
@rvernica I've built my local SciDB 19.11.5 as `RelWithDebInfo` and see the tests complete in about 2 minutes.  Built as `Debug`, the tests complete in 13 minutes.  Are we running a release build of SciDB in the docker image or a debug build?

**Comment by rvernica** [comment]
I'm running the release from the `rpm`/`deb` packages. From the Docker images:
```
# scidb --version
SciDB Version: 19.11.5
Build Type: RelWithDebInfo
Commit: f8334b6
Copyright (C) 2008-2019 SciDB, Inc.
```
Without `remove` it takes about five minutes to complete on my system. With `remove` eight threads finish, one fails, and one gets stuck waiting and never finishes.

**Comment by dgosselin-p4** [comment]
Thank you for the feedback.  I'll keep digging on my end.

**Comment by dgosselin-p4** [comment]
@rvernica is there a chance that the docker image has a shim built from the shim's v19.3 branch?  `shim -v` doesn't show the shim commit (although it has a place for it) and the version of SciDB parrots whatever's installed there.  Locally, if I run the v19.3 shim against SciDB 19.11.5, then in the `pytest -n1` case, the test is half-as-fast in the case where it removes the arrays.

**Comment by rvernica** [comment]
Good point about the Shim version. The Docker containers don't have Git installed, so the Shim version is not recorded when Shim is built. I added Git to the Docker containers and the new builds will have the Shim version.

Now, regarding the Shim branch used, it works as follows:
1. Docker containers currently [pull](https://github.com/rvernica/docker-library/blob/master/scidb/19.11/Dockerfile.xenial#L9) a specific Shim [version](https://github.com/Paradigm4/shim/commit/16fe182411be3192030e6d8f587969fc875cabf0). They compile and [install](https://github.com/rvernica/docker-library/blob/master/scidb/19.11/Dockerfile.xenial#L105) it in the container.
1. In Travis, we always get the [branch being built](https://github.com/Paradigm4/shim/blob/master/.travis.yml#L39) (`master`, PR, release, etc.) We compile, [install](https://github.com/Paradigm4/shim/blob/master/.travis.yml#L57), and [test](https://github.com/Paradigm4/shim/blob/master/.travis.yml#L68) it. So, the Travis test will *not* use the Shim version already installed in the container.

You can test a different Shim version by doing `git checkout` in the `~/shim` directory in the AWS instance. That directory is mounted on the Docker container. So, once you checkout the version you want, you can go in the container in `/shim` and build and test that specific version.

**Comment by dgosselin-p4** [comment]
Looks like what's in the container, both installed and running, are commit sha 15d4e37b50e9823a8fd707b884895e7693d3974e.  I took an `md5sum` of both the shim executable in `/shim` and at `/opt/scidb/19.11/bin` to match them, with the assumption that the one in `/shim` (a git repo) was built from `/shim`.  Long story short, my theory about mismatched shim versions doesn't hold.

**Comment by dgosselin-p4** [comment]
@rvernica would you mind deploying another AWS instance that is identical to what we have for 19.11 but has 19.3 running instead?  IIRC this issue doesn't occur running against 19.3, so it would be nice to compare various logs from both systems to get a better understanding of what's happening.

**Comment by rvernica** [comment]
Done. Instance name is `Rares Worker 19.3`. It has the same settings.

**Comment by dgosselin-p4** [comment]
Thanks!  I renamed the other system to `Rares Worker 19.11`.

**Comment by dgosselin-p4** [comment]
@rvernica I'm setting this aside for a bit to get some other work done and will return next week.  Please do not remove the AWS instances that you created for me as I will need them.  Thanks.

**Comment by ksens** [comment]
@dgosselin-p4 @rvernica 

Just checking:

What is the status w.r.t. this issue? Due to this issue (and potentially related issue https://github.com/rvernica/docker-library/issues/12), I am unable to use my usual development environment (using scidb on Docker). 

I have workarounds but would like to start using my usual dev setup whenever possible. Thanks in advance.

**Comment by rvernica** [comment]
Travis runs the Shim tests on weekly basis. I've been watching it for the last few weeks and it consistently succeeded. Nothing changed in the Shim code. I guess that we are not seeing failures because of some underlying changes in the Travis platform. That is, Travis now uses more powerful instances to run the Shim tests.

**Comment by ksens** [comment]
Thanks. Got a workaround for now for my local Docker setup -- https://github.com/rvernica/docker-library/issues/12#issuecomment-647435930.
